### ISSUE TITLE
Matrices do not perform a deep copy (inconsistent with vector)

### ISSUE THREAD

**Issue by chrispepper1989** [dwmkerr/glmnet#1]
There is some inconsistencies with GLMnet in the way it treats matrices and vertices.

vec2 bob = new vec2(1,1);
vec2 fred = bob;  // fred is now a copy of bob
fred = fred \* new vec(2,2,); //bob is still 1,1

mat4 bob= mat4.identity();
mat4 fred = bob; //both bob and fred point to the same internal array

glm.translate(fred , by); //bob & fred have now changed

I suggest that you overload the "=" operator so that matrices perform a deep copy. Or provide a clone function like:

```
    public static mat4 CloneMatrix(mat4 matrix)
    {
        vec4[] cols = new[]
        {
            matrix[0],
            matrix[1],
            matrix[2],
            matrix[3]

        };
        return new mat4(cols);
    }
```

**Comment by chrispepper1989** [comment]
Another thing that needs to be considered is "identity" 

i.e. 

myMat = mat4.identity();

At the moment that effectively changes the pointer of myMat to the pointer of "mat4.identity()" (i think) 

Which if im not mistaken is a "new" so we don't change the identity matrix. But it does mean that if mat is left as a clonable reference users are going to experience inconsistent behavior e.g.:

mat4 camera;

mat4 GetCamera()
{
      return camera; //users holding a ref to this would expect it to change on "UpdateCamera"
}

UpdateCamera()
{
    camera = mat4.identy() //now "camera" is pointing to something else 
    //make changes to camera

    //user with "camera" ref wonders why its not updated
}

**Comment by dwmkerr** [comment]
Yes it's certainly non-trivial to get this right, I think we'll have to document a consistent approach and describe it clearly in the readme as something which deviates from the behaviour of the C++ GLM
